### ISSUE TITLE
Windows 11 with Secure Boot doesn't work

### ISSUE THREAD

**Issue by CBRocker** [FOGProject/fogproject#597]
Hi,

Facts:
FOG Version 1.5.10
FOG Dev. Version: 1.5.10.30

First please give me Infos, how I can register me at the FOG Forum.
I fill out the fields username and password, confirm the captcha code, after click on "Jetzt registrieren" I see the message
"Captcha not verified, are you a robot?" No matter what I type, the error always comes up.

Well, the primary issue:

We have some Dell Inspiron 5420 All-in-One-PCs with Windows 11. We have problems with the Secure Boot/UEFI.
PXE and WOL are set in the BIOS.

Enabled Secure Boot, the PC boots from the installed Windows 11.
Disabled Secure Boot, the PC boots with PXE, the FOG/iPXE 1.21.1+ starts (looking for new Images and so on).
But at the FOG Project screen (where is the text "Host is registered as ...") (see below)
at the end of the line "Boot form hard disk" the counter counts down from 3 to 0, but there is a endless loop
and it doesn't go any further. 

![Image](https://github.com/FOGProject/fogproject/assets/96444759/56e1b804-2464-4815-97de-885ac576060c)

Enabled Secure Boot, the PXE/deploying Image doesn't work.

How can it solved?
FOG works with Windows 11/Secure Boot?

Thx for ur help

**Comment by mastacontrola** [comment]
FOG Does not work out of the box with secure boot nor has it ever.

Disable secure boot,
capture your image, (or deploy)
enable secure boot.

If you require secure boot consistently, there are methods available to do it but it's not a foolproof method I doubt.

https://forums.fogproject.org/topic/13832/secureboot-issues/8?_=1721229080146
https://forums.fogproject.org/post/145434

**Comment by darksidemilk** [comment]
I just re-read this and realized you're disabling secure boot and not being able to boot to disk, not so much trying to boot to fog with secure boot.
We do have a fix in the 1.6 beta (update from the working-1.6 branch instead of the master branch to try it out)
In the latest versions of ipxe they made it possible to boot to a uefi disk with the sanboot command.
We updated the boot to hard disk behavior to use this new functionality and included the latest version of ipxe in working 1.6. you can see the sanboot changes in fog here:
https://github.com/FOGProject/fogproject/blob/b9dcbe397aed0f03aa27f1308f0066d91dc13f04/packages/web/lib/fog/bootmenu.class.php#L177

You could also follow the instructions here https://docs.fogproject.org/en/latest/compile_ipxe_binaries to compile the latest ipxe for your current version and edit the equivalent sanboot line manually.
You also want to change your efi exit option from refind to sanboot for this to work. 

That's a rough guideline on what needs doing, let me know if you need more detailed help, I just don't have the time right this second to dig into deeper detail

**Comment by darksidemilk** [comment]
@CBRocker were you able to try the newer ipxe and sanboot option?

**Comment by sfonteneau** [comment]
Hi team, for your information, Broadcom's ipxe works in secure boot:

https://knowledge.broadcom.com/external/article/280113/updated-64bit-ipxeefi-ipxe-v1211-binarie.html

you just have to match the url with that:

http://{nextserver}:4433/Altiris/iPXE/GetPxeScript.aspx

The URL must chain to the ipxe script :)

I haven't tried it with fog (I'm using wimboot) but I thought it might be useful to you!

**Comment by darksidemilk** [comment]
@sfonteneau I've actually see that same article, well actually it looks like it's been updated as recently as last month. When I first found it the ipxe version compiled was from before ipxe added the ability to use sanboot in uefi mode, so I didn't bother.
I suppose one could make an apache vhost on that 4433 port and create dummy folders and a symlink file for `/Altiris/iPXE/GetPxeScript.aspx` that points to `fog/service/ipxe/boot.php`
worth a shot.
I did just throw it in tftpboot and tried to make an autoexe.ipxe by copying default.ipxe to see if it would then just skip the built-in script to no avail, but it did boot with secureboot enabled.

**Comment by darksidemilk** [comment]
<img width="568" height="67" alt="Image" src="https://github.com/user-attachments/assets/901f9d8c-db55-40cc-87e3-0947f7d6d501" />
darn, didn't work. Certainly something to mess with though.

**Comment by sfonteneau** [comment]
4433 is http and not https?

**Comment by darksidemilk** [comment]
Yes, I actually was setting it up as https because it follows a typical https port pattern, but they did http and I made sure my redirect to boot.php worked. But default.ipxe adds params to boot.php so that's probably part of what would be missing, so maybe it has to chain to default.ipxe instead or something. If I get more time to mess with it I will.

**Comment by sfonteneau** [comment]
I haven't tried it with fog but I would have done it like that:

```
<VirtualHost *:80 *:4433>
    ProxyRequests Off
    ProxyPreserveHost On

    ProxyPassMatch "^/Altiris/iPXE/GetPxeScript\.aspx$" "http://127.0.0.1/fog/service/ipxe/boot.php"
    ProxyPassReverse "/Altiris/iPXE/GetPxeScript.aspx" "http://127.0.0.1/fog/service/ipxe/boot.php"

</VirtualHost>
```

**Comment by darksidemilk** [comment]
Sadly the broadcom altiris ipxe file appears to not have been built with the PARAM_CMD option, so our default.ipxe script that uses the params command to send the host info to boot.php doesn't work.
I redirected the aspx to `/tftpboot/default.ipxe` which will look something like this format:

```
#!ipxe
set arch ${buildarch}
iseq ${arch} i386 && cpuid --ext 29 && set arch x86_64 ||
params
param mac0 ${net0/mac}
param arch ${arch}
param platform ${platform}
param product ${product}
param manufacturer ${product}
param ipxever ${version}
param filename ${filename}
param sysuuid ${uuid}
isset ${net1/mac} && param mac1 ${net1/mac} || goto bootme
isset ${net2/mac} && param mac2 ${net2/mac} || goto bootme
:bootme
chain http://192.168.0.254/fog/service/ipxe/boot.php##params
```

but we get this with the signed file

<img width="308" height="52" alt="Image" src="https://github.com/user-attachments/assets/c09607e3-dbc4-42d5-8f8b-b63464533afc" />

So I tried making a new `default-params.ipxe` file in /tftpboot, and symlinked the expected file to that, which got closer to the desired result.
I did it like this

```
#!ipxe
set arch ${buildarch}
set fogip 192.168.0.254
iseq ${arch} i386 && cpuid --ext 29 && set arch x86_64 ||
isset ${net1/mac} || goto bootme
isset ${net2/mac} || goto bootme2macs
isset ${net3/mac} || goto bootme3macs
:bootme4macs
chain http://${fogip}/fog/service/ipxe/boot.php?mac=${net0/mac}&mac0=${net1/mac}&mac1=${net2/mac}&mac2=${net3/mac}&arch=${arch}&platform=${platform}&product=${product}&manufacturer=${manufacturer}&ipxever=${version}&filename=${filename}&sysuuid=${uuid}
:bootme3macs
chain http://${fogip}/fog/service/ipxe/boot.php?mac=${net0/mac}&mac0=${net1/mac}&mac1=${net2/mac}&arch=${arch}&platform=${platform}&product=${product}&manufacturer=${manufacturer}&ipxever=${version}&filename=${filename}&sysuuid=${uuid}
:bootme2macs
chain http://${fogip}/fog/service/ipxe/boot.php?mac=${net0/mac}&mac0=${net1/mac}&arch=${arch}&platform=${platform}&product=${product}&manufacturer=${manufacturer}&ipxever=${version}&filename=${filename}&sysuuid=${uuid}
:bootme
chain http://${fogip}/fog/service/ipxe/boot.php?mac=${net0/mac}&arch=${arch}&platform=${platform}&product=${product}&manufacturer=${manufacturer}&ipxever=${version}&filename=${filename}&sysuuid=${uuid}
```

This gets things much farther as it then boots to the menu and it does show the correct hostname, but I don't have my custom (nor the default) background image, just a blue screen.
I believe that boot.php is generating the pxe menu script and that will also be using the missing params command so it will continue to fail unless we also re-write that. 
I did try to make it go to fog compatibility mode which would load fos and test if it will load in the secure boot chain. But sadly that through an error instead of downloading and loading init.xz. It also failed to boot to the next disk with sanboot. I think we're also using params to find that disk but I don't remember.

<img width="624" height="207" alt="Image" src="https://github.com/user-attachments/assets/4f4c340c-36c3-4302-911d-d0b2db56555f" />

Maybe with more work it could be made to work and then you could do secure boot without having to create your own keys and add them to each systems efi and then compile and sign your own ipxe binary. Or Broadcom could just compile their file with the PARAM_CMD option and then it would just work when redirected to our default.ipxe file I imagine.

Someone that's done more work on the ipxe scripts and chain in fog would know more. 

At this point though, we've gone far from the original issue here and should probably make a separate issue or forum post for this.

**Comment by sfonteneau** [comment]
I think you've put your finger on the problem.:

https://github.com/search?q=repo%3AFOGProject%2Ffogproject+php%23%23params&type=code

in my case (waptserver),  I don't use "param", and it works.

**Comment by baerus** [comment]
Hi there, I am way too unfamiliar (technically) with the details, but on a broader level: Will FOG try to support secure boot imaging (capture/deploy) in the future? It becomes quite annoying to disable and enable it each time when I need to deploy to my 13 PCs, as I go in the UEFI by hand for each machine each time (maybe there's a smarter way as well?..)

I also read this: https://forums.fogproject.org/topic/13832/secureboot-issues and it looked somewhat promising. But again technically I am too far away to understand the details. Before I want to ruin my marriage and dig down that hole, I want to know: It's possible, it's worthwhile, it's (kinda) supported etc. I search for solutions which also work in the next 5 years etc.

Maybe one of you guys can enlighten me, how is FOG planning to proceed? Secure Boot becomes more relevant each year (I come from the gaming environment, where more and more games requires secure boot) and I wonder how FOGs position is/will be on that: "not our problem", "we have something" or "we're working on something in x years".
Thanks!
(If this is the wrong place for such a question, and somewhat it feels like it, please link me to somewhere better if possible)

**Comment by darksidemilk** [comment]
It's something we very much want to do, but Microsoft doesn't make it easy or free to get it done. I mean it shouldn't be too easy or it wouldn't be secure. Ultimately, for a Linux based and internally customized system like fog, the best solution of the moment is getting your computers to trust an internal or fog generated cert that is the root for the secure boot keys for the pxe boot file, background image, bzImage and init.xz etc (everything needs to be signed by a trusted cert). There's a forum post with a rough guide on how to make that happen but we hope to have a simpler implementation of it. It probably won't be built in, but will eventually be easy to add, it'll just be a while before any of the devs have time to work it out. Something like a one time setup of secure boot keys in the pxe menu when you register a host then you turn secure boot on after that and your good. But again it'll be a bit. There is a turnkey something fork of fog where they are using secure boot by reusing the canonical cert and switching from pxe to grub for the boot menu, but they also switched fos to be alpine based. If I remember when I get to a computer I'll add links. I would hope within 5 years we'll get it worked out, and it's very much something we want to do, it's just finding the best way that fits the most use cases and requires the least maintenance

**Comment by abotzung** [comment]
Hi,

I'm the developer of _FOGUefi_, a fork of FOS that allows the use of FOG on PCs with Secure Boot enabled.

As @DarkSideMilk points out, signing the bootloader and Linux kernel is a lengthy and complicated process. 

Therefore, the project is based on shim, GRUB, the Linux kernel, and signed modules from Ubuntu Noble (to validate/pass the boot process) and Alpine Linux as the distribution.

Note that iPXE is not modified and can be used alongside GRUB, only the boot file (in the DHCP server / option 67) will need to be changed to "**shimx64.efi**".

If you're interested, the project is here: https://github.com/abotzung/foguefi

**Comment by guldnerthomas** [comment]
Hi, I’ve noticed iPXE v2.0 is out with native Secure Boot support. I'm not sure if this could finally simplify things for FOG Project, especially regarding the boot file signature. What do you think about its potential impact on FOG?

**Comment by sfonteneau** [comment]
Interesting! https://github.com/ipxe/ipxe/releases/latest/download/ipxeboot.tar.gz contains an x86_64-sb directory (sb as in Secure Boot), so if we boot using shimx64.efi, it should work. thanks.

**Comment by darksidemilk** [comment]
We still need the fos kernel executable to be signed for all to work with secure boot.

**Comment by RomsnRamos** [comment]
> Hi,
> 
> I'm the developer of _FOGUefi_, a fork of FOS that allows the use of FOG on PCs with Secure Boot enabled.
> 
> As [@darksidemilk](https://github.com/darksidemilk) points out, signing the bootloader and Linux kernel is a lengthy and complicated process.
> 
> Therefore, the project is based on shim, GRUB, the Linux kernel, and signed modules from Ubuntu Noble (to validate/pass the boot process) and Alpine Linux as the distribution.
> 
> Note that iPXE is not modified and can be used alongside GRUB, only the boot file (in the DHCP server / option 67) will need to be changed to "**shimx64.efi**".
> 
> If you're interested, the project is here: https://github.com/abotzung/foguefi


Hello abotzung,

First of all, thank you for your hard work. Since I see that issues can't be created in the GitHub repository, I'm posting this here hoping you'll see it. I've successfully implemented foguefi, and although I encountered two problems that I was able to solve with Gemini—one being that some symbolic links for Apache were missing, and the other requiring me to modify something in the grrubmenu.class.php file to force it to download the kernel and initrd via HTTP (because it was always doing it via TFTP, which took about 12 minutes)—the error I'm getting now is when I try to deploy an image. When it tries to dump the image, it throws an error saying that the kernel parameters are missing. However, the task on the FOG server is created, and when it restarts from the network, it detects that there's a task and then deploys the image correctly. It's not crucial, but it would be great if it worked.

Regards @abotzung 

<img width="911" height="741" alt="Image" src="https://github.com/user-attachments/assets/060ccf5a-b110-4560-b5b9-636b1bd6214f" />

**Comment by mastacontrola** [comment]
Cross-linking: the Secure Boot half of this now has a concrete analysis in #957.

Short version of what changed since mastacontrola's reply — his answer ("FOG does not work out of the box with secure boot nor has it ever") is still correct today, but the ground has moved. iPXE 2.0.0 (March 2026) ships UEFI Secure Boot support via a Microsoft-signed shim, and FOG is now pinned to it.

What I found looking at what it would take for FOG specifically:

- The **iPXE half is close**. FOG's EFI builds already enable the `shim` command, the installer already picks a boot file per architecture off DHCP option 93, and iPXE 2.0 can pull its boot script from `autoexec.ipxe` over TFTP, so FOG would not need a custom binary at all.
- The **gate is the FOS kernel**. A Secure Boot build of iPXE compiles out the Linux/bzImage loader entirely, and the EFI path it would use instead hands the image to the firmware's `LoadImage()` — so FOG's unsigned FOS kernel is rejected by firmware policy, then rejected again by the shim. Making that work means FOG signing its own kernels, which means FOG owning a signing key.

So this is not blocked on a missing feature any more, it is blocked on a decision. Details and the source references are in #957.

The other half of this thread — booting to a UEFI disk with Secure Boot *disabled*, which darksidemilk correctly re-read this as — is already addressed in 1.6 via `sanboot`.

**Comment by mastacontrola** [comment]
We now have methodologies that I think can close this thank you!
